File: src/main/java/com/baidu/algorithm/backtracking/_216_Combination_Sum_III.py
class Solution(object):
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(res, [], k, 9, 0, n)

        return res

    def dfs(self, res, cur_res, k, cur, cur_sum, n):
        if k == 0:
            if cur_sum == n:
                res.append(cur_res[:])
        elif cur_sum >= n or cur == 0:
            return
        else:
            self.dfs(res, cur_res, k, cur - 1, cur_sum, n)

            cur_res.append(cur)
            self.dfs(res, cur_res, k - 1, cur - 1, cur_sum + cur, n)
            cur_res.pop()

File: src/main/java/com/baidu/algorithm/backtracking/_216_Combination_Sum_III.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.backtracking;

import java.util.ArrayList;
import java.util.List;

/**
 * _216_Combination_Sum_III
 *
 * @author xuhaoran01
 */
public class _216_Combination_Sum_III {
    public void dfs(List<List<Integer>> res, List<Integer> curRes, int cur, int k, int n) {
        if (k == 0) {
            if (n == 0) {
                res.add(new ArrayList<>(curRes));
            }

            return;
        } else if (n <= 0 || cur > 9) {
            return;
        }

        dfs(res, curRes, cur + 1, k, n);

        curRes.add(cur);
        dfs(res, curRes, cur + 1, k - 1, n - cur);
        curRes.remove(curRes.size() - 1);
    }

    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(res, new ArrayList<>(), 1, k, n);

        return res;
    }
}
